### DISCUSSION TITLE
How should unsupported custom types be handled?

### DISCUSSION THREAD

**Discussion by Tornquist** [prisma/prisma#9541] [Q&A] [answered]
__Summary:__

I am struggling to handle raw queries with unsupported types. I've found that I have to explicitly map to `::text` to get the data to return without throwing errors in the rust layer.

__Details:__

I have a custom postgres type in a table as follows:

```sql
CREATE TYPE public.message_frequency AS (
  message text,
  frequency bigint
);

CREATE TABLE public.data (
  id uuid NOT NULL,
  entries public.message_frequency[] NOT NULL
)
```

As-expected, when I build the initial prisma schema, it came in as `Unsupported('_message_frequency')[]`. __I understand that these fields will not be available in the prisma client, and that I'll have to use `$queryRaw`.__

I'm surprised that `$queryRaw` cannot handle fetching this data as strings or any and just returning the raw data to me.

```js
const data = await prisma.$queryRawUnsafe(`
  SELECT * FROM data where id = '${id}'
`)
```

and

```js
const data = await prisma.$queryRawUnsafe(`
  SELECT id, data FROM data where id = '${id}'
`)
```

both fail with

```js
PrismaClientKnownRequestError:
Invalid `prisma.queryRaw()` invocation:

  Raw query failed. Code: `N/A`. Message: `error deserializing column 2: cannot convert between the Rust type `core::option::Option<alloc::vec::Vec<alloc::string::String>>` and the Postgres type `_message_frequency``
    at cb (/<...>/node_modules/@prisma/client/runtime/index.js:36952:17) {
  code: 'P2010',
  clientVersion: '3.1.1',
  meta: {
    code: 'N/A',
    message: 'error deserializing column 30: cannot convert between the Rust type `core::option::Option<alloc::vec::Vec<alloc::string::String>>` and the Postgres type `_message_frequency`'
  }
}
```

I've tried removing the entry from the schema and rebuilding the client, but it still fails to map the value with the error above. The presence of that data type blocks direct sql access without explicitly mapping the type like:

```js
const data = await prisma.$queryRawUnsafe(`
  SELECT id, entries::text FROM data where id = '${id}'
`)
```

I'm fine to parse the type in JS directly, but I'd like to avoid situations where if someone uses a `*` in a raw query they'll generate a type that crashes the execution.

Can I force that mapping to text in the prisma layer?

**Comment by ryands17** [comment, marked as answer]
@Tornquist 👋 

For unsupported types, you can perform operations like add, edit, delete but you cannot fetch that data back as Prisma cannot serialise that data. For all the operations you would need to use raw queries at the moment. I don't think there's a system that prevents using a `select *`, but there was an issue related to this that you can add your comments on: https://github.com/prisma/prisma/issues/9353

**Reply by Tornquist** [reply]
Ah! Okay, thank you. I wasn't aware of the specific limitations on the raw query types/actions.

So it sounds like mapping to text explicitly and then handling it myself is the best/only way to use prisma to read data with those types existing on that table?

**Reply by ryands17** [reply]
Yup that's how you would need to go about it for the time being.

**Reply by Tornquist** [reply]
Perfect, thanks!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
